Labelling test updates

delhomer opened the pull request:
This PR refactors some test related to feature detection and semantic segmentation labelling.

Before there were three tests for each problem, that were quite redundant. Now, there are only two tests per problem:
- `test_problem_concise`, which tests the case where the array `a` contains exactly the values stored in `labels`
- `test_problem_sparse`, which tests the case where the array `a` values and `labels` are not the same (it corresponds to the case where there are some non-evaluated labels).

This PR fix issue #52.

delhomer committed:
tests: fix bincount-related exception in test_semantic_segmentation_labelling fix #52

delhomer committed:
tests: update feature detection labelling tests

Remove random array generation, and delete a duplicated tests. There
are now `test_feature_detection_labelling_concise`, which tests the
case where labels and array values are the same; and
`test_feature_detection_labelling_sparse` which tests the case where
the array contains values that are not stored in reference label list.

delhomer committed:
tests: update the semantic segmentation labelling tests

This commit produces `test_semantic_segmentation_labelling_concise`
and `test_semantic_segmentation_labelling_sparse`, on the model of
feature detection tests.

garaud commented on tests/test_generators.py:
```diff
@@ -8,58 +8,60 @@
+                   [2, 1, 0, 0]]])
+    labels = np.unique(a).tolist()
+    MIN, MAX = np.amin(a), np.amax(a)
+    with pytest.raises(AttributeError):
```
I didn't see that for my first review when you updated some test, but this shouldn't be an `AttributeError` exception which should be raised. It's better if we raise a `ValueError` exception. We have to update the `feature_detection_labelling` function.

delhomer replied:
The fact is that this portion of code raises an `AttributeError` ! If we replace it with a `ValueError`, the test does not pass any more.

garaud replied:
That's why i wrote "we have to update the `feature_detection_labelling` function" in that way.

garaud commented on tests/test_generators.py:
```diff
@@ -8,58 +8,60 @@
-                          [True, False, True]]
+        b = generator.feature_detection_labelling(a, ['0', '1', '2'])
+    b = generator.feature_detection_labelling(a, labels)
+    assert len(labels) != range(np.amin(a), np.amax(a))
```
Is it not strange to compare an integer to a generator here? This is always True anyway when you do `2 != range(0,3)`.

delhomer replied:
Oups, little forget. It is not `range(...)`, but `len(range(...)`. I will fix this.

garaud replied:
Just (np.max(a) - np.min(a)) + 1 instead of len(range())

delhomer replied:
Ok, modified.

delhomer replied:
By the way, it is `np.amax(a)-np.amin(a)`, instead of `np.amax(a)-np.amin(a)+1`.

delhomer replied:
Hum, my bad, it is `np.amax(a)-np.amin(a)+1`. Modified.

garaud reviewed (changes requested):
I'm not very familiar with the "concise" or "sparse" vocabulary. "sparse" makes me think to a sparsed array/matrix in linear algebra.

I'm not other propositions for renaming this for now. Let's keep that.

delhomer committed:
tests: fix a wrong assertion in test_generator

Currently we check if an integer is different from a range. Of course
the assertion is successful. To be accurate, the assertion must
consider an integer instead of a range, *i.e.*
`np.amax(a)-np.amin(a)+1`.

delhomer committed:
tests: modify an error type when executing feature_detection_labelling and semantic_segmentation_labelling

Move the AttributeError and replace it with ValueError, as the
exception arises when considering a list of non-integer values.
